feat: introduce a new site info API endpoint

dcshzj opened the pull request:
## Problem

<!-- What problem are you trying to solve? What issue does this close? -->

There are currently no existing endpoints suitable for the site dashboard to use to obtain some basic information about the site.

## Solution

<!-- How did you solve the problem? -->

**Breaking Changes**

<!-- Does this PR contain any backward incompatible changes? If so, what are they and should there be special considerations for release? -->

- [ ] Yes - this PR contains breaking changes
- [x] No - this PR is backwards compatible

**Features**:

- A new site info API endpoint has been introduced, which presents:
    - The email address of the user that last committed to the staging and master branches
    - The date that the latest commit to the staging and master branches were made
    - The staging URL of the site
    - The live production site URL
- A new `getUrlsOfSite()` function has been created based on a refactoring of the existing `getStagingUrl()` function.
- A new `getCommitAuthorEmail()` function has been created to retrieve the email address of the user that made a particular commit.
- The existing `getStagingUrl()` function has been refactored to utilise the new method of obtaining the staging URL.

## Tests

<!-- What tests should be run to confirm functionality? -->

**Unit tests**: Run `npm run test`

seaerchin commented on src/services/identity/SitesService.ts:
```diff
@@ -78,6 +84,107 @@ class SitesService {
+    }
+
+    const deployment = site?.deployment
+    const output: SiteUrls = {
+      staging: deployment?.stagingUrl ? deployment.stagingUrl : "",
+      prod: deployment?.productionUrl ? deployment.productionUrl : "",
+    }
```
note that `Deployment` has `allowNull: false` set on both of these so we probably shouldn't do this; instead, what we could do is to define a `doesDeploymentExists` variable that determines if we should fetch using `configYmlService` (or just `!deployment`) 

this does make it slightly less future proof (eg: we want to `allowNull`) though... lmk what you think!

dcshzj replied:
If deployment (or site) does not exist, then we will just continue which is intended.

seaerchin commented on src/services/identity/SitesService.ts:
```diff
@@ -78,6 +84,107 @@ class SitesService {
+    // 2. From the config.yml file
+    // 3. From the GitHub repository description
+    // Otherwise, returns an empty string for the url
+    const { siteName } = sessionData
```
i think the code below is pretty janky to read and we should instead write it as a flow of something like

1. obtain data 
2. use util function to assign properties if it exists in data 
3. check if both properties are valid otherwise do 1-3 again

wdyt about this?

dcshzj replied:
Adjusted in 4432732.

seaerchin commented on src/services/identity/SitesService.ts:
```diff
@@ -78,6 +84,107 @@ class SitesService {
+      }
+    }
+
+    return output
```
note that `output` here is not guaranteed to be well-formed. legacy sites could potentially have misbehaving behaviour and hence, still have undefined `staging | prod`. 

at the end, we could validate properties and return either 1. well formed output, 2. a single output (either `staging | prod` only or 3. error

double check wrt github repo desc @alexanderleegs

dcshzj replied:
Adjusted in 4432732. If all 3 sources of data does not contain the info we need, then we will return an UnprocessableError.

seaerchin commented on src/services/identity/SitesService.ts:
```diff
@@ -154,29 +261,13 @@ class SitesService {
-        (token) => token.includes("http") && token.includes("staging")
+    const { staging } = await this.getUrlsOfSite(sessionData, ["staging"])
+    if (!staging) {
+      throw new NotFoundError(
+        `${sessionData.siteName} does not have a staging url`
```
shouldn't throw, prefer returning an error and let the router decide what to do

dcshzj replied:
Changed in ed61f55.

seaerchin commented on src/services/identity/SitesService.ts:
```diff
@@ -194,6 +285,41 @@ class SitesService {
+    const stagingCommit = ((await this.gitHubService.getLatestCommitOfBranch(
+      sessionData,
+      "staging"
+    )) as unknown) as GitHubCommitData
```
this is a dangerous cast. we should validate the properties and use an assertion.

dcshzj replied:
Added an assertion function in 9ca0838.

seaerchin commented on src/services/identity/SitesService.ts:
```diff
@@ -194,6 +285,41 @@ class SitesService {
+    const prodCommit = ((await this.gitHubService.getLatestCommitOfBranch(
+      sessionData,
+      "master"
+    )) as unknown) as GitHubCommitData
```
same

dcshzj replied:
9ca0838

alexanderleegs commented on src/routes/v2/authenticated/sites.ts:
```diff
@@ -84,6 +84,25 @@ export class SitesRouter {
+  > = async (req, res) => {
+    const { userSessionData } = res.locals
+    const { siteName } = req.params
+    const userWithSiteSessionData = this.addSiteNameToSessionData(
+      userSessionData,
+      siteName
+    )
```
Actually we should already have siteName in the sessionData here! Via the extra attachSiteHandler middleware for this route

dcshzj replied:
Removed in a12ad20!

alexanderleegs commented on src/services/identity/SitesService.ts:
```diff
@@ -78,6 +84,107 @@ class SitesService {
+    let configYmlData: ConfigYmlData = {}
+    let repoDescTokens: string[] = []
+
+    const site = await this.siteRepository.findOne({
+      where: { name: siteName },
+      include: {
+        model: Deployment,
+        as: "deployment",
+      },
+    })
```
Actually, given that this is the SitesService itself, does this suggestion mean that we would need to query SitesService for the existence of the site, before we can call this endpoint? Also, this might cause duplicated code if we want to reuse this method (e.g. if multiple methods call `getUrlsOfSite`, both would need to first query some generic `SitesService.get` method to retrieve a `Site`, before this method could be called) - I'm not sure if this behaviour is desirable

dcshzj replied:
Discussed offline. This is a refactor that we will do separately soon<sup>TM</sup>

dcshzj commented on src/fixtures/identity.ts:
```diff
@@ -138,3 +140,23 @@ export const mockSiteOrmResponseWithNoCollaborators = {
+  message: mockCommitMessage2,
+  fileName: mockCommitFileName2,
+  userId: mockIsomerUserId,
+}
```
Updated in b126c53.

dcshzj commented on src/routes/v2/authenticated/sites.ts:
```diff
@@ -84,6 +84,25 @@ export class SitesRouter {
+      userWithSiteSessionData
+    )
+    return res.status(200).json(siteInfo)
+  }
```
Good point! Added in e9b0112.

dcshzj commented on src/services/db/GitHubService.js:
```diff
@@ -284,6 +284,27 @@ class GitHubService {
 
+  async getLatestCommitOfBranch(sessionData, branch) {
+    const { accessToken } = sessionData
+    const { siteName } = sessionData
```
Whoops, combined in d335ebf.

dcshzj commented on src/services/identity/SitesService.ts:
```diff
@@ -78,6 +84,107 @@ class SitesService {
+
+    // Commit message created as part of phase 2 identity
+    if (message.startsWith("{") && message.endsWith("}")) {
+      const { userId }: { userId: string } = JSON.parse(message)
```
TIL about the JSONParseError, added a `try/catch` in 5ef57a4.

dcshzj commented on src/services/identity/SitesService.ts:
```diff
@@ -78,6 +84,107 @@ class SitesService {
+      }
+    }
+
+    const deployment = site?.deployment
```
Adjusted in 66defaa.

dcshzj commented on src/services/db/GitHubService.js:
```diff
@@ -284,6 +284,27 @@ class GitHubService {
     return { treeSha, currentCommitSha }
   }
 
+  async getLatestCommitOfBranch(sessionData, branch) {
```
Discussed offline, would prefer to centralise everything so that when refactoring to TypeScript, it is easier to find all the functions that we need to refactor in one place.

dcshzj commented on src/services/identity/SitesService.ts:
```diff
@@ -78,6 +84,107 @@ class SitesService {
+
+      if (user) {
+        return user.email
+      }
```
This branch will not be entered and will be handled later, as the GitHubCommitData is guaranteed to have the email address, which will be returned.

dcshzj commented on src/services/identity/SitesService.ts:
```diff
@@ -78,6 +84,107 @@ class SitesService {
+        staging: "",
+        prod: "",
+      }
+    }
```
Adjusted in 4432732.

dcshzj commented on src/services/identity/SitesService.ts:
```diff
@@ -78,6 +84,107 @@ class SitesService {
+
+    for (const urlType of urlTypes) {
+      // We only call ConfigYmlService only when necessary
+      if (!output[urlType] && Object.keys(configYmlData).length === 0) {
```
Adjusted in 4432732.

dcshzj commented on src/services/identity/SitesService.ts:
```diff
@@ -78,6 +84,107 @@ class SitesService {
+      }
+
+      // We only call GitHubService only when necessary
+      if (!output[urlType] && repoDescTokens.length == 0) {
```
Adjusted in 4432732.

dcshzj commented on src/services/identity/SitesService.ts:
```diff
@@ -78,6 +84,107 @@ class SitesService {
+
+  async getUrlsOfSite(
+    sessionData: UserWithSiteSessionData,
+    urlTypes: SiteUrlTypes[]
```
Adjusted in 4432732.

dcshzj commented:
Added a fix where the endpoint for getting the latest commit is incorrect.

Also added a stub for getting the email address of the user that last published the site, which should be obtained from the review request table.

seaerchin commented on src/services/identity/SitesService.ts:
```diff
@@ -66,6 +67,89 @@ class SitesService {
+    )
+
+    // Only replace the urls if they are not already present
+    siteUrls = {
+      staging:
+        configYmlData.staging && !siteUrls.staging
+          ? configYmlData.staging
+          : siteUrls.staging,
+      prod:
+        configYmlData.prod && !siteUrls.prod
+          ? configYmlData.prod
+          : siteUrls.prod,
+    }
```
could we have a helper method for this? (cos also done below but minor)

also, we should create a new object rather than passing `siteUrls` - this modifies `siteUrl` in place and js doesn't pass by value so this is potentially dangerous

dcshzj replied:
Hmm I think it might be clearer here instead of having a separate helper method?

Also adjusted the name/assignment in 3692b08.

seaerchin commented on src/services/identity/SitesService.ts:
```diff
@@ -66,6 +67,89 @@ class SitesService {
+    )
+
+    // Only replace the urls if they are not already present
+    siteUrls = {
```
see above

dcshzj replied:
Done in 3692b08.

seaerchin commented on src/services/identity/SitesService.ts:
```diff
@@ -89,11 +173,15 @@ class SitesService {
 
-      if (user) {
-        return user.email
+        if (user) {
+          return user.email
```
could we only return if `user.email` is a string? this is `string | null | undefined`, and we can do better by returning the commit author's email

dcshzj replied:
Good catch! Fixed in fdf07a5.

seaerchin commented on src/services/identity/SitesService.ts:
```diff
@@ -104,18 +192,32 @@ class SitesService {
-    urlTypes: SiteUrlTypes[]
-  ): Promise<SiteUrls> {
+    sessionData: UserWithSiteSessionData
+  ): Promise<SiteUrls | NotFoundError> {
```
specialise the error here to give more context

seaerchin commented on src/services/identity/SitesService.ts:
```diff
@@ -125,64 +227,24 @@ class SitesService {
-      prod: deployment?.productionUrl ? deployment.productionUrl : "",
+    // Note: site may be null if the site does not exist
+    let siteUrls: SiteUrls = {
+      staging: site?.deployment?.stagingUrl ? site.deployment.stagingUrl : "",
+      prod: site?.deployment?.productionUrl
+        ? site.deployment.productionUrl
+        : "",
```
we can use `??` or `||` - both easier to read

dcshzj replied:
Improved in 42e73df.

seaerchin commented on src/services/identity/SitesService.ts:
```diff
@@ -125,64 +227,24 @@ class SitesService {
-        // repo descriptions have varying formats, so we look for the first link
-        repoDescTokens = description.replace("/;/g", " ").split(" ")
-      }
+    siteUrls = await this.insertUrlsFromConfigYml(siteUrls, sessionData)
+    siteUrls = await this.insertUrlsFromGitHubDescription(siteUrls, sessionData)
```
should aim to merge - lodash has a method for this iirc

dcshzj replied:
Improved in 42e73df.

seaerchin commented on src/services/identity/SitesService.ts:
```diff
@@ -260,13 +322,18 @@ class SitesService {
         `${sessionData.siteName} does not have a staging url`
       )
     }
+
```
actually we could do a hierarchy of errors like so: `NotFoundError` -> `Site...` -> `StagingUrl...`

dcshzj commented on src/routes/v2/authenticated/sites.ts:
```diff
@@ -70,17 +60,17 @@ export class SitesRouter {
+    // Check for error and throw
+    if (stagingUrl instanceof BaseIsomerError) {
+      throw stagingUrl
+    }
```
Done in 99d9140.

dcshzj commented on src/routes/v2/authenticated/sites.ts:
```diff
@@ -70,17 +60,17 @@ export class SitesRouter {
-      siteName
-    )
+    const { userWithSiteSessionData } = res.locals
     const stagingUrl = await this.sitesService.getStagingUrl(
```
Done in 99d9140.

dcshzj commented on src/routes/v2/authenticated/sites.ts:
```diff
@@ -89,17 +79,18 @@ export class SitesRouter {
+    // Check for error and throw
+    if (siteInfo instanceof BaseIsomerError) {
+      throw siteInfo
+    }
```
99d9140

dcshzj commented on src/services/db/__tests__/GitHubService.spec.js:
```diff
@@ -557,6 +558,51 @@ describe("Github Service", () => {
+      })
+    })
+
+    it("Getting an invalid branch should throw UnprocessableError", async () => {
```
Good point! Added in cdf94ea.

dcshzj commented on src/services/identity/SitesService.ts:
```diff
@@ -66,6 +67,89 @@ class SitesService {
     this.tokenStore = tokenStore
   }
 
+  assertGitHubCommitData(commit: any): commit is GitHubCommitData {
```
Adjusted in 1eb32dd.

dcshzj commented on src/services/identity/SitesService.ts:
```diff
@@ -66,6 +67,89 @@ class SitesService {
 
+  assertGitHubCommitData(commit: any): commit is GitHubCommitData {
+    if (
+      commit &&
+      (commit as GitHubCommitData).author !== undefined &&
+      (commit as GitHubCommitData).author.name !== undefined &&
+      (commit as GitHubCommitData).author.date !== undefined &&
+      (commit as GitHubCommitData).author.email !== undefined &&
+      (commit as GitHubCommitData).message !== undefined
```
Whoops, adjusted in 1eb32dd.

seaerchin reviewed (approved):
lgtm
